File: 01_goBase/src/leedcode/o21_08/o08_25_字符串转换整数/01_answer.go
package main

import (
	"math"
	"unicode"
)

// 定义 状态
const (
	START = iota
	SIGN
	IN_NUMBER
	END
)

// 状态下对应变化状态
// 0 ' '
// 1 +/-
// 2 数字
// 3 其他
var statusMap = map[int][4]int{
	START:     {START, SIGN, IN_NUMBER, END},
	SIGN:      {END, END, IN_NUMBER, END},
	IN_NUMBER: {END, END, IN_NUMBER, END},
	END:       {END, END, END, END},
}

// 存储结果
type Automaton struct {
	ans    int64
	sign   int64
	status int
}

// 计算并终止
// @return true 终止
func (maton *Automaton) calc(char rune) bool {
	// 更新当前状态
	maton.updateStatus(char)
	// 判断当前状态
	switch maton.status {
	case SIGN:
		// 计算符号
		if char == '-' {
			maton.sign = -1
		} else {
			maton.sign = 1
		}
	case IN_NUMBER:
		// 添加数字
		maton.ans = maton.ans*10 + int64(char) - int64('0')
		if maton.sign > 0 {
			if maton.ans > math.MaxInt32 {
				maton.ans = math.MaxInt32
				return true
			}
		} else {
			if maton.ans > -math.MinInt32 {
				maton.ans = -math.MinInt32
				return true
			}
		}
	case END:
		return true
	}
	return false
}

// 更新状态
func (maton *Automaton) updateStatus(char rune) {
	var idx int
	if char == ' ' {
		idx = 0
	} else if char == '+' || char == '-' {
		idx = 1
	} else if unicode.IsDigit(char) {
		idx = 2
	} else {
		idx = 3
	}
	maton.status = statusMap[maton.status][idx]
}

// 更新状态
func (maton *Automaton) getAns() int {
	return int(maton.ans * maton.sign)
}
func myAtoi(str string) int {
	// 初始化 变量
	automaton := Automaton{status: START, sign: 1, ans: 0}
	for _, char := range str {
		if automaton.calc(char) {
			return automaton.getAns()
		}
	}
	return automaton.getAns()
}

File: 01_goBase/src/leedcode/o21_08/o08_25_字符串转换整数/01_answer.py
from enum import Enum


# 定义状态枚举
class Status(Enum):
	START = 0
	SIGN = 1
	IN_NUMBER = 2
	END = 3


# 定义最大值
INT_MAX = 2 ** 31 - 1
INT_MIN_DE = 2 ** 31


class Automaton:
	def __init__(self):
		# 符号
		self.sign = 1
		# 大小
		self.ans = 0
		# 状态
		self.status = Status.START
		# 状态map
		self.table = {
			Status.START: [Status.START, Status.SIGN, Status.IN_NUMBER, Status.END],
			Status.SIGN: [Status.END, Status.END, Status.IN_NUMBER, Status.END],
			Status.IN_NUMBER: [Status.END, Status.END, Status.IN_NUMBER, Status.END],
			Status.END: [Status.END, Status.END, Status.END, Status.END],
		}

	def calc(self, char: str) -> bool:
		self.updateStatus(char)
		if self.status == Status.SIGN:
			if char == '-':
				self.sign = -1
		elif self.status == Status.IN_NUMBER:
			# 计算内容
			self.ans = self.ans * 10 + int(char)
			if self.sign < 0:
				if self.ans > INT_MIN_DE:
					self.ans = INT_MIN_DE
					return True
			else:
				if self.ans > INT_MAX:
					self.ans = INT_MAX
					return True
		return False
	def get_ans(self) -> int:
		return self.ans * self.sign

	def updateStatus(self, char: str):
		if char.isspace():
			status = Status.START
		elif char == '+' or char == '-':
			status = Status.START
		elif char.isdigit():
			status = Status.START
		else:
			status = Status.END
		self.status = self.table[self.status][status]


class Solution:
	def myAtoi(self, s: str) -> int:
		maton = Automaton()
		for item in s:
			if maton.calc(item):
				return maton.getAns()
		return maton.getAns()


maton = Automaton()
print(maton)
print(maton.get_ans())